core/src/main/antlr4/com/databricks/labs/remorph/parsers/snowflake/SnowflakeParser.g4

```diff
@@ -3699,34 +3699,43 @@ expr
     | expr over_clause
     | cast_expr
     | expr COLON_COLON data_type // Cast also
-    | try_cast_expr
     | json_literal
     | trim_expression
     | function_call
-    | subquery
-    | expr IS null_not_null
-    | expr NOT? IN L_PAREN (subquery | expr_list) R_PAREN
-    | expr NOT? ( LIKE | ILIKE) expr (ESCAPE expr)?
-    | expr NOT? RLIKE expr
-    | expr NOT? (LIKE | ILIKE) ANY L_PAREN expr (COMMA expr)* R_PAREN (ESCAPE expr)?
+    | subquery // Probably wrong
+    | expr predicate_partial
     | primitive_expression //Should be latest rule as it's nearly a catch all
     ;
 
+predicate_partial
+    : IS null_not_null
+    | NOT? IN L_PAREN (subquery | expr_list) R_PAREN
+    | NOT? ( LIKE | ILIKE) expr (ESCAPE expr)?
+    | NOT? RLIKE expr
+    | NOT? (LIKE | ILIKE) ANY L_PAREN expr (COMMA expr)* R_PAREN (ESCAPE expr)?
+    | NOT? BETWEEN expr AND expr
+    ;
+
+json_path
+    : json_path_elem (DOT json_path_elem)*
+    ;
+
+json_path_elem
+    : ID | DOUBLE_QUOTE_ID
+    ;
+
```

Comment:
We should not need partial rules to avoid self recursion. The rule should look something like this:

```antlr
searchCondition
    : LPAREN searchCondition RPAREN         #scPrec
    | NOT searchCondition                   #scNot
    | searchCondition AND searchCondition   #scAnd
    | searchCondition OR searchCondition    #scOr
    | predicate                             #scPred
    ;

predicate
    : EXISTS LPAREN subquery RPAREN
    | freetextPredicate
    | expression comparisonOperator expression
    | expression ME expression ////SQL-82 syntax for left outer joins; PE. See https://stackoverflow.com/questions/40665/in-sybase-sql
    | expression comparisonOperator (ALL | SOME | ANY) LPAREN subquery RPAREN
    | expression NOT* BETWEEN expression AND expression
    | expression NOT* IN LPAREN (subquery | expressionList) RPAREN
    | expression NOT* LIKE expression (ESCAPE expression)?
    | expression IS nullNotnull
    ;
```

Which is what TSQL looks like right now (though it will likely change)